model/task/generate_tasks_estimations.go

```diff
@@ -0,0 +1,118 @@
+package task
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/utility"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+const (
+	lookBackTime = 7 * 24 * time.Hour // one week
+)
+
+type generateTasksEstimationsResults struct {
+	DisplayName        string  `bson:"_id"`
+	EstimatedCreated   float64 `bson:"est_created"`
+	EstimatedActivated float64 `bson:"est_activated"`
+}
+
+func (t *Task) setGenerateTasksEstimations() error {
+	// Do not run if estimations have already been cached.
+	if t.EstimatedNumGeneratedTasks != nil && t.EstimatedNumActivatedGeneratedTasks != nil {
+		return nil
+	}
+
+	match := bson.M{
+		BuildVariantKey:   t.BuildVariant,
+		ProjectKey:        t.Project,
+		DisplayNameKey:    t.DisplayName,
+		GeneratedTasksKey: true,
+		StatusKey: bson.M{
+			"$in": evergreen.TaskCompletedStatuses,
+		},
+		StartTimeKey: bson.M{
+			"$gt": time.Now().Add(-1 * lookBackTime),
+		},
+		FinishTimeKey: bson.M{
+			"$lte": time.Now(),
+		},
+	}
+
+	pipeline := []bson.M{
+		{
+			"$match": match,
+		},
+		{
+			"$project": bson.M{
+				DisplayNameKey:                1,
+				NumGeneratedTasksKey:          1,
+				NumActivatedGeneratedTasksKey: 1,
+				IdKey:                         0,
+			},
+		},
+		{
+			"$group": bson.M{
+				"_id": fmt.Sprintf("$%s", DisplayNameKey),
+				"est_created": bson.M{
+					"$avg": fmt.Sprintf("$%s", NumGeneratedTasksKey),
+				},
+				"est_activated": bson.M{
+					"$avg": fmt.Sprintf("$%s", NumActivatedGeneratedTasksKey),
+				},
+			},
+		},
+	}
```

Comment:
If this query is going to be like the estimated task duration query, I believe it may have to also cache in a way that's similar to the estimated duration, or else this will be too expensive. I believe the estimated duration logic _used_ to just fetch the duration for every task, but it became too expensive, so it now caches results.

If I understand, the way the task estimated duration is a little more efficient is that it cuts down on the frequency of repeated queries. It first [queries the estimated duration](https://github.com/evergreen-ci/evergreen/blob/4a4a164a29a534f4f43e81310b2f44c846ffe19d/model/task/task.go#L3390) and then caches that result for some amount of time for that task name. Once that TTL has passed, it'll get the latest estimate by querying again for that task.